### PR TITLE
MESMER-X: Test `distrib_cov`

### PR DESCRIPTION
**Pull request by veni-vidi-vici-dormivi**
- [x] Tests added
 - [x] Fully documented, including `CHANGELOG.rst`

### CODE REVIEW

File: mesmer/mesmer_x/train_l_distrib_mesmerx.py [hunk 1]
```diff
@@ -540,21 +544,19 @@ def __init__(
...
-        if self.method_fit in [
-            "dogleg",
-            "trust-ncg",
-            "trust-krylov",
-            "trust-exact",
-            "COBYLA",
-            "SLSQP",
-            "CG",
-            "Newton-CG",
+        if self.method_fit not in [
```
Comment (mathause) [inline mesmer/mesmer_x/train_l_distrib_mesmerx.py:553]: nice

File: mesmer/mesmer_x/train_l_distrib_mesmerx.py [hunk 2]
```diff
@@ -614,8 +616,8 @@ def __init__(
         ):
             raise ValueError(
                 "Lack of consistency on the options 'type_fun_optim',"
-                " 'threshold_stopping_rule' and 'ind_year_thres', not sure if the"
-                " stopping rule will be employed"
+                " 'threshold_stopping_rule' and 'ind_year_thres', threshold_stopping_rule",
+                "and 'ind_year_thres' must be used together, and only for 'fcNLL'",
```
Comment (mathause) [inline mesmer/mesmer_x/train_l_distrib_mesmerx.py:620]:
```suggestion
                " 'threshold_stopping_rule' and 'ind_year_thres', threshold_stopping_rule,"
                " and 'ind_year_thres' must be used together, and only for 'fcNLL'",
```

File: mesmer/mesmer_x/train_l_distrib_mesmerx.py [hunk 3]
```diff
@@ -649,7 +651,7 @@ def _get_weights_nll(self, n_bins_density=40):
         # interpolating over whole region
         gmt_hist, edges = np.histogramdd(sample=tmp, bins=bins.T)
 
-        gmt_bins_center = [0.5 * (edge[1:] + edges[:-1]) for edge in edges]
+        gmt_bins_center = [0.5 * (edge[1:] + edge[:-1]) for edge in edges]
```
Comment (mathause) [inline mesmer/mesmer_x/train_l_distrib_mesmerx.py:654]: nice find

File: mesmer/mesmer_x/train_l_distrib_mesmerx.py [hunk 4]
```diff
@@ -38,6 +38,7 @@ def wrapper(*args, **kwargs):
     return wrapper
 
 
+# TODO: would want to switch this, have a distrib class that takes xarrays and have a training func that potentially works on xarrays
```
Comment (mathause) [inline mesmer/mesmer_x/train_l_distrib_mesmerx.py:41]:
```suggestion
# TODO: enable distrib class and training func for xarray objs
```

File: tests/unit/test_mesmer_x_distrib_cov.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+
+    expression = Expression("norm(loc=c1 * __tas__, scale=c2)", expr_name="exp1")
+
+    dist = distrib_cov(targ, {"tas": pred}, expression)
+
+    np.testing.assert_equal(dist.data_targ, targ)
+    np.testing.assert_equal(dist.data_pred, {"tas": pred})
+    np.testing.assert_equal(dist.weights_driver, np.ones(n) / n)
+    assert dist.n_sample == n
+    assert dist.expr_fit == expression
```
Comment (mathause) [inline tests/unit/test_mesmer_x_distrib_cov.py:21]: Does that actually work? Ah it does but I think it checks that it's the same objects ant thus better to use `is` to be more explicit.
```suggestion
    assert dist.expr_fit is expression
```

File: tests/unit/test_mesmer_x_distrib_cov.py [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+    assert not dist.add_test
+    assert dist.data_targ_addtest is None
+    assert dist.data_preds_addtest is None
+    assert dist.threshold_min_proba == 1e-09
+    assert dist.boundaries_params == expression.boundaries_parameters
+    assert dist.boundaries_coeffs == {}
+    assert dist.first_guess is None
+    assert dist.func_first_guess is None
+    assert dist.n_coeffs == 2
```
Comment (mathause) [inline tests/unit/test_mesmer_x_distrib_cov.py:30]: I think that should be a property of `expr_fit` (no need to do that here, though)

File: tests/unit/test_mesmer_x_distrib_cov.py [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+    n = 250
+    pred = np.linspace(0, 1, n)
+    targ = rng.normal(loc=2 * pred, scale=0.1, size=n)
+
+    expression = Expression("norm(loc=c1 * __tas__, scale=c2)", expr_name="exp1")
+
+    data_targ_addtest = rng.normal(loc=2 * pred, scale=0.1, size=n)
+    data_preds_addtest = {"tas": np.linspace(0, 0.9, n)}
+    threshold_min_proba = 0.1
+    boundaries_params = {"loc": [-10, 10], "scale": [0, 1]}
```
Comment (mathause) [inline tests/unit/test_mesmer_x_distrib_cov.py:59]:
```suggestion
    boundaries_params = {"loc": [-10, 10], "scale": [-1, 1]}
```

Could try -1 here to check we end up with 0 again

File: tests/unit/test_mesmer_x_distrib_cov.py [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+
+    np.testing.assert_equal(dist.data_targ, targ)
+    np.testing.assert_equal(dist.data_pred, {"tas": pred})
+    np.testing.assert_equal(dist.weights_driver, dist.get_weights())
+    # np.testing.assert_equal(dist.weights_driver, dist._get_weights_nll()) # WHY NOT???
+    np.testing.assert_equal(dist.first_guess, first_guess)
+    np.testing.assert_equal(dist.data_targ_addtest, data_targ_addtest)
+    np.testing.assert_equal(dist.data_preds_addtest, data_preds_addtest)
+    assert dist.n_sample == n
+    assert dist.expr_fit == expression
```
Comment (mathause) [inline tests/unit/test_mesmer_x_distrib_cov.py:105]:
```suggestion
    assert dist.expr_fit is expression
```

File: tests/unit/test_mesmer_x_distrib_cov.py [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+            expression,
+            first_guess=np.array([1, 2, 3]),
+        )
+
+    with pytest.raises(ValueError, match="`options_solver` must be a dictionary"):
+        distrib_cov(
+            np.array([1, 2, 3]),
+            {"tas": np.array([1, 2, 3])},
+            expression,
+            options_solver="this is not a dictionary",
```
Comment (mathause) [inline tests/unit/test_mesmer_x_distrib_cov.py:197]: :smile:

**Review by mathause — APPROVED** [review]
Thanks that looks good.

**Comment by veni-vidi-vici-dormivi** [comment]
Ahh sorry, this is not at all done yet, I just pushed it to get the coverage report 😅

File: tests/unit/test_mesmer_x_distrib_cov.py (hunk 5 above)
Comment (veni-vidi-vici-dormivi) [reply tests/unit/test_mesmer_x_distrib_cov.py:21]: Ah thanks!

**Comment by mathause** [comment]
> Ahh sorry, this is not at all done yet, I just pushed it to get the coverage report 😅

Ahh well looks good already :wink:

**Comment by veni-vidi-vici-dormivi** [comment]
@mathause added two tests for the weighting but only very basic ones. I am merging without another review and we can change something afterwards to if we want to.

**Pull request merged**